Speculative fix for crash mitigation in JSModuleRecord::instantiateDeclarations

hyjorc1 opened the pull request:
#### 79f077301b073479201ec6d29ac564fe2c6738e4
<pre>
Speculative fix for crash mitigation in JSModuleRecord::instantiateDeclarations
<a href="https://bugs.webkit.org/show_bug.cgi?id=255013">https://bugs.webkit.org/show_bug.cgi?id=255013</a>
rdar://107634974

Reviewed by Mark Lam.

This is a speculative fix to mitigate a crash in JSModuleRecord::instantiateDeclarations
which is due to an undefined imported module. However, the existing implementation
should already handle the undefined case by throwing and catching an exception when
encountering an undefined JSValue. This patch ensures that the target pointer will not
have the undefined value. If it does, we&apos;ll throw an exception.

* Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
(JSC::JSModuleRecord::instantiateDeclarations):

Canonical link: <a href="https://commits.webkit.org/262608@main">https://commits.webkit.org/262608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e93705205374092f2dc1b7e68d6e1e044c9340d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2034 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2062 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2957 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2096 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2140 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2103 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1860 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2051 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1822 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2807 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1811 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1801 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1710 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1699 "Build is in progress. Recent messages:") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1868 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1849 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2978 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1938 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1866 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2092 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1792 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1812 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/488 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/503 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1974 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2138 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/603 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/JSCJSValue.h:
```diff
@@ -353,6 +353,8 @@ class JSValue {
     JSCell* asCell() const;
 
+    void* asRawPointer() const;
+
```
Undo this.  You don't need this.

MenloDorian commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -400,7 +400,12 @@ inline double JSValue::asDouble() const
+{
+    return reinterpret_cast<void*>(u.asBits.payload);
+}
+
```
Undo this.

MenloDorian commented on Source/JavaScriptCore/runtime/JSCJSValueInlines.h:
```diff
@@ -600,6 +605,11 @@ inline bool JSValue::isNumber() const
+{
+    return u.ptr;
+}
+
```
Undo this.

MenloDorian commented on Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
```diff
@@ -151,6 +151,18 @@ void JSModuleRecord::instantiateDeclarations(JSGlobalObject* globalObject, Modul
+
+        // rdar://107531050: Speculative crash mitigation
+        // TODO: rdar://107634974
+        if (UNLIKELY(importedModule == bitwise_cast<AbstractModuleRecord*>(jsUndefined().asRawPointer()))) {
```
Instead of `jsUndefined().asRawPointer()`, use `encodedJSUndefined()`.

MenloDorian reviewed (approved):
r=me with fixes.

hyjorc1 committed:
Speculative fix for crash mitigation in JSModuleRecord::instantiateDeclarations
https://bugs.webkit.org/show_bug.cgi?id=255013
rdar://107634974

Reviewed by Mark Lam.

This is a speculative fix to mitigate a crash in JSModuleRecord::instantiateDeclarations
which is due to an undefined imported module. However, the existing implementation
should already handle the undefined case by throwing and catching an exception when
encountering an undefined JSValue. This patch ensures that the target pointer will not
have the undefined value. If it does, we'll throw an exception.

* Source/JavaScriptCore/runtime/JSModuleRecord.cpp:
(JSC::JSModuleRecord::instantiateDeclarations):

Canonical link: https://commits.webkit.org/262608@main
